File: library/src/main/java/com/owncloud/android/lib/resources/e2ee/ToggleEncryptionRemoteOperation.java
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2017-2024 Nextcloud GmbH and Nextcloud contributors
 * SPDX-FileCopyrightText: 2017 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-License-Identifier: MIT
 */
package com.owncloud.android.lib.resources.e2ee;

import com.nextcloud.common.SessionTimeOut;
import com.nextcloud.common.SessionTimeOutKt;
import com.owncloud.android.lib.common.OwnCloudClient;
import com.owncloud.android.lib.common.operations.RemoteOperation;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.lib.common.utils.Log_OC;
import com.owncloud.android.lib.resources.files.ReadFolderRemoteOperation;

import org.apache.commons.httpclient.HttpMethodBase;
import org.apache.commons.httpclient.HttpStatus;
import org.apache.commons.httpclient.methods.DeleteMethod;
import org.apache.commons.httpclient.methods.PutMethod;


/**
 * Set encryption of a folder
 */

public class ToggleEncryptionRemoteOperation extends RemoteOperation<Void> {

    private static final String TAG = ToggleEncryptionRemoteOperation.class.getSimpleName();
    private static final String ENCRYPTED_URL = "/ocs/v2.php/apps/end_to_end_encryption/api/v1/encrypted/";

    private final long localId;
    private final String remotePath;
    private final boolean encryption;

    private final SessionTimeOut sessionTimeOut;

    /**
     * Constructor
     */
    public ToggleEncryptionRemoteOperation(long localId, String remotePath, boolean encryption) {
        this.localId = localId;
        this.remotePath = remotePath;
        this.encryption = encryption;
        sessionTimeOut = SessionTimeOutKt.getDefaultSessionTimeOut();
    }

    public ToggleEncryptionRemoteOperation(long localId, String remotePath, boolean encryption, SessionTimeOut sessionTimeOut) {
        this.localId = localId;
        this.remotePath = remotePath;
        this.encryption = encryption;
        this.sessionTimeOut = sessionTimeOut;
    }

    /**
     * @param client Client object
     */
    @Override
    protected RemoteOperationResult<Void> run(OwnCloudClient client) {
        RemoteOperationResult<Void> result;
        HttpMethodBase method = null;

        ReadFolderRemoteOperation remoteFolderOperation = new ReadFolderRemoteOperation(remotePath);
        RemoteOperationResult remoteFolderOperationResult = remoteFolderOperation.execute(client);

        // Abort if not empty
        // Result has always the folder and maybe children, so size == 1 is ok
        if (remoteFolderOperationResult.isSuccess() && remoteFolderOperationResult.getData().size() > 1) {
            return new RemoteOperationResult<>(false, "Non empty", HttpStatus.SC_FORBIDDEN);
        }

        try {
            String url = client.getBaseUri() + ENCRYPTED_URL + localId;
            if (encryption) {
                method = new PutMethod(url);
            } else {
                method = new DeleteMethod(url);
            }

            // remote request
            method.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE);
            method.addRequestHeader(CONTENT_TYPE, FORM_URLENCODED);

            int status = client.executeMethod(method, sessionTimeOut.getReadTimeOut(), sessionTimeOut.getConnectionTimeOut());

            if (status == HttpStatus.SC_OK) {
                result = new RemoteOperationResult<>(true, method);
            } else {
                result = new RemoteOperationResult<>(false, method);
                client.exhaustResponse(method.getResponseBodyAsStream());
            }
        } catch (Exception e) {
            result = new RemoteOperationResult<>(e);
            Log_OC.e(TAG, "Setting encryption status of " + localId + " failed: " + result.getLogMessage(),
                     result.getException());
        } finally {
            if (method != null) {
                method.releaseConnection();
            }
        }
        return result;
    }
}

File: library/src/main/java/com/owncloud/android/lib/resources/e2ee/ToggleEncryptionRemoteOperation.kt
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-License-Identifier: MIT
 */
package com.owncloud.android.lib.resources.e2ee

import com.nextcloud.common.NextcloudClient
import com.nextcloud.common.OkHttpMethodBase
import com.nextcloud.common.SessionTimeOut
import com.nextcloud.common.defaultSessionTimeOut
import com.nextcloud.operations.DeleteMethod
import com.nextcloud.operations.PutMethod
import com.owncloud.android.lib.common.operations.RemoteOperation
import com.owncloud.android.lib.common.operations.RemoteOperationResult
import com.owncloud.android.lib.common.utils.Log_OC
import com.owncloud.android.lib.resources.files.ReadFolderRemoteOperation
import okhttp3.RequestBody.Companion.toRequestBody
import org.apache.commons.httpclient.HttpStatus

class ToggleEncryptionRemoteOperation
    @JvmOverloads
    constructor(
        private val localId: Long,
        private val remotePath: String?,
        private val encryption: Boolean,
        private val sessionTimeOut: SessionTimeOut = defaultSessionTimeOut
    ) : RemoteOperation<Unit>() {
        @Suppress("TooGenericExceptionCaught")
        override fun run(client: NextcloudClient): RemoteOperationResult<Unit> {
            val folderResult = ReadFolderRemoteOperation(remotePath.orEmpty()).execute(client)
            if (folderResult.isSuccess && folderResult.resultData.size > 1) {
                return RemoteOperationResult(false, "Non empty", HttpStatus.SC_FORBIDDEN)
            }

            return runCatching { toggleEncryption(client) }
                .getOrElse { e ->
                    Log_OC.e(TAG, "Setting encryption status of $localId failed", e)
                    RemoteOperationResult(e as? Exception ?: RuntimeException(e))
                }
        }

        private fun toggleEncryption(client: NextcloudClient): RemoteOperationResult<Unit> {
            val (status, method) = executeWithFallback(client)

            return RemoteOperationResult<Unit>(status == HttpStatus.SC_OK, method).also {
                method.releaseConnection()
            }
        }

        /**
         * Tries the v2 endpoint first, falling back to v1 on 404/500.
         * Returns the final status code and the method used.
         */
        private fun executeWithFallback(client: NextcloudClient): Pair<Int, OkHttpMethodBase> {
            val timedClient = client.withSessionTimeOut(sessionTimeOut)

            val v2Method = buildMethod(client, ENCRYPTED_URL_V2)
            val v2Status = timedClient.execute(v2Method)

            val needsFallback = v2Status == HttpStatus.SC_NOT_FOUND || v2Status == HttpStatus.SC_INTERNAL_SERVER_ERROR
            if (!needsFallback) return v2Status to v2Method

            v2Method.releaseConnection()
            val v1Method = buildMethod(client, ENCRYPTED_URL_V1)
            val v1Status = timedClient.execute(v1Method)
            return v1Status to v1Method
        }

        private fun buildMethod(
            client: NextcloudClient,
            baseUrl: String
        ): OkHttpMethodBase {
            val uri = "${client.baseUri}$baseUrl$localId"

            val method =
                if (encryption) {
                    PutMethod(uri, false, "".toRequestBody(null))
                } else {
                    DeleteMethod(uri, false)
                }

            method.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE)
            method.addRequestHeader(CONTENT_TYPE, FORM_URLENCODED)

            return method
        }

        companion object {
            private val TAG = ToggleEncryptionRemoteOperation::class.java.simpleName
            private const val ENCRYPTED_URL_V1 = "/ocs/v2.php/apps/end_to_end_encryption/api/v1/encrypted/"
            private const val ENCRYPTED_URL_V2 = "/ocs/v2.php/apps/end_to_end_encryption/api/v2/encrypted/"
        }
    }
